test/gpu/native/studies/shoc/sort.chpl

```diff
@@ -0,0 +1,679 @@
+    const findBlocks    = findGlobalWorkSize    / SCAN_BLOCK_SIZE;
+    const reorderBlocks = reorderGlobalWorkSize / SCAN_BLOCK_SIZE;
+
+    // radixSortBlocks Kernel here
+    radixSortBlocks(radixGlobalWorkSize, nbits, startbit, tempKeys, tempValues, keys, values);
+
+    // findRadixOffset Kernel here
+    // Use offsets again
+    findRadixOffsets(findGlobalWorkSize, tempKeys, counters, blockOffsets, startbit, numElements, findBlocks);
+
+    scanArrayRecursive(countersSum, counters, 16*reorderBlocks, 0, scanBlockSums);
+
+    // reorderData Kernel here
+    reorderData(reorderGlobalWorkSize, startbit, keys, values, tempKeys, tempValues,
+         blockOffsets, countersSum, counters, reorderBlocks);
+}
+
+proc scanArrayRecursive(ref outArray : [] uint(32), ref inArray : [] uint(32),
+                        numElements, level, ref blockSums : [] list(uint(32))){
+    // Kernels hadle 8 elements per thread
+    const numBlocks : uint(32) = max(1,ceil(numElements:real
+        / (4.0:real * SCAN_BLOCK_SIZE)) : uint(32)) : uint(32);
+
+    const sharedEltsPerBlock : uint(32) = (SCAN_BLOCK_SIZE * 2) : uint(32);
+    const sharedMemSize : uint(32) = numBytes(uint(32)) * sharedEltsPerBlock;
+
+    const fullBlock : bool = (numElements == numBlocks * 4 * SCAN_BLOCK_SIZE);
+
+    // CUDA version uses dim3 for no apparent reason because they
+    // created dim3 with only x populated for grid and block size
+
+    //execute scan
+    if(numBlocks>1){
+        // Scan Kernel here
+        var t = blockSums[level].toArray();
+        scanKernel(numBlocks, outArray, inArray, t, numElements, fullBlock, true);
+        blockSums[level] = new list(t);
+    } else {
+        // Different Scan Kernel here
+        var t = blockSums[level].toArray();
+        scanKernel(numBlocks, outArray, inArray, t, numElements, fullBlock, false);
+        blockSums[level] = new list(t);
+    }
+
+    if(numBlocks > 1){
+        var t = blockSums[level].toArray();
+        scanArrayRecursive(t, t, numBlocks, level+1, blockSums);
+        blockSums[level] = new list(t);
+        // VectorAddUniform4 Kernel Here
+        t = blockSums[level].toArray();
+        vectorAddUniform4(outArray, t, numElements);
+        blockSums[level] = new list(t);
+    }
+}
+
+proc verifySort(ref keys : [] uint(32), ref values : [] uint(32), const size){
+    var passed = true;
+    for i in 0..<size-1 {
+        if(keys[i] > keys[i+1]){
+            passed = false;
+            writeln("Fail at ", i, ": ", keys[i], " > ", keys[i+1]);
+            break;
+        }
+    }
+    write("Test ");
+    if(passed) then writeln("Passed");
+    else writeln("Failed");
+    return passed;
+}
+
+// /////////////////////////////////////////////////////////////////////////////////
+// // Kernels
+// ////////////////////////////////////////////////////////////////////////////////
+
+
+proc radixSortBlocks(radixGlobalWorkSize, const nbits : uint(32), const startbit : uint(32),
```

Comment:
Nitpicky: But just to make it easier to see the correspondence between the CUDA and Chapel version I'd put these functions in the same order as they appear in the CUDA version:

(https://github.com/vetter/shoc/blob/master/src/cuda/level1/sort/sort_kernel.cu)

In other words: scanLSB, radixSortBlocks, findRadixOffsets, reorderData, scanLocalMem, scan(Kernel), vectorAddUniform4

Reply:
I put it in the following order:
scanLSB, scan4, adixSortBlocks, findRadixOffsets, reorderData, scanLocalMem, scan(Kernel), vectorAddUniform4